### PR TITLE
Add a slack streamalert app

### PR DESCRIPTION
**Pull request by GarretReece**
to: @airbnb/streamalert-maintainers
cc: @jacknagz 
size: medium

## Background

StreamAlert currently doesn't have an app that collects Slack logs.

## Changes

* Adds an app that will collect logs from Slack. Supported log types are access logs and integrations logs.
* Schema definitions for the two new logs types are included.
* Unit tests for the app as well as schema validation tests are included.
* Documentation for authorizing the app is included in the README in the apps directory.

## Testing

Unit tests and schema validation tests included. App has been deployed to my test environment and works.

### CODE REVIEW

**Comment by jacknagz** [comment]
Hey @GarretReece, we'll review soon 😄

File: app_integrations/apps/slack.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        data = self._get_request_data()
+
+        success, response = self._make_post_request(
+            url, headers, data, False)
+
+        if not success:
+            LOGGER.exception('Received bad response from slack')
+            return False
+
+        if not u'ok' in response.keys() or not response[u'ok']:
```
Comment (ryandeivert) [inline app_integrations/apps/slack.py]: can you omit the unicode `u` prefix here (and elsewhere). technically it's fine, just overly verbose.

also, to be more consistent with elsewhere, the can this please be changed to:

`if 'ok' not in response or not response['ok']`...

or: `if not response.get('ok')` would even do

File: app_integrations/apps/slack.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+            LOGGER.exception('Received error or warning from slack')
+            return False
+
+        self._more_to_poll = self._check_for_more_to_poll(response)
+
+        results = self._filter_response_entries(response)
+
+        if not self._more_to_poll:
+            self._last_timestamp = int(time.time())
```
Comment (ryandeivert) [inline app_integrations/apps/slack.py]: This value can be updated even if there the `_more_to_poll` prop is true. Updating it here will make sure it gets updated in the saved stated, and will prevent missing or duplicative data in the event of a function timeout.

File: app_integrations/apps/slack.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    def _filter_response_entries(self, response):
+        if 'logs' in response:
+            return [x for x in response['logs'] if int(x['date']) > self._last_timestamp]
+        return False
+
+    @classmethod
+    def _sleep_seconds(cls):
+        """Return the number of seconds this polling function should sleep for
+        between requests to avoid failed requests. The Slack team.integrationLog API
+        has Tier 2 limiting, which is 20 requests per minute.
```
Comment (ryandeivert) [inline app_integrations/apps/slack.py]: good find!!

File: app_integrations/apps/slack.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+        if (response['paging']['page'] >= self._SLACK_API_MAX_PAGE_COUNT and
+                response['paging']['count'] >= self._SLACK_API_MAX_ENTRY_COUNT):
+            self._before_time = response['logins'][-1]['date_first']
+            self._next_page = 1
+            return True
+
+        self._next_page += 1
+        return response['paging']['pages'] > response['paging']['page']
+
+
```
Comment (ryandeivert) [inline app_integrations/apps/slack.py]: nitpick- there's an extra line here

File: app_integrations/apps/slack.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    def _endpoint(cls):
+        return cls._SLACK_ACCESS_LOGS_ENDPOINT
+
+    def _filter_response_entries(self, response):
+        if 'logins' in response:
+            return [x for x in response['logins'] if x['date_last'] > self._last_timestamp]
+        return False
+
+    def _check_for_more_to_poll(self, response):
+        '''if we hit the maximum possible number of returned entries, there may still be more
```
Comment (ryandeivert) [inline app_integrations/apps/slack.py]: Please use double-quotes (`"""`) to surround docstrings

File: app_integrations/apps/slack.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        return not ('paging' in response.keys() and
+                    response['paging']['pages'] == response['paging']['page'])
+
+    def _filter_response_entries(self, response):
+        """The slack endpoints don't provide a programatic way to filter for new results,
+        so subclasses must implement their own endpoint specific filtering methods"""
+        raise NotImplementedError("Subclasses must implement the _filter_response_entries method")
+
+    def _get_request_data(self):
+        '''The Slack API takes additional parameters to its endpoints in the body of the request.
```
Comment (ryandeivert) [inline app_integrations/apps/slack.py]: Please use double-quotes (`"""`) to surround docstrings

File: app_integrations/apps/slack.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+                }
+            }
+
+    def _check_for_more_to_poll(self, response):
+        self._next_page += 1
+        return not ('paging' in response.keys() and
+                    response['paging']['pages'] == response['paging']['page'])
+
+    def _filter_response_entries(self, response):
+        """The slack endpoints don't provide a programatic way to filter for new results,
```
Comment (ryandeivert) [inline app_integrations/apps/slack.py]: spelling: `programatic` --> `programmatic`

File: app_integrations/apps/slack.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+    @classmethod
+    def _sleep_seconds(cls):
+        raise NotImplementedError('Subclasses should implement the _sleep_seconds method')
+
+    @classmethod
+    def service(cls):
+        return 'slack'
+
+    @classmethod
+    def date_formatter(cls):
```
Comment (ryandeivert) [inline app_integrations/apps/slack.py]: If not special formatting is needed, you should be able to omit this method altogether (it's not an abstractmethod that is required to be implemented)

File: tests/unit/app_integrations/test_apps/test_slack.py [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+        self._app = SlackAccessApp(AppConfig(get_valid_config_dict('slack')))
+
+    def test_sleep_seconds(self):
+        """SlackAccessApp - Sleep Seconds"""
+        assert_equal(3, self._app._sleep_seconds())
+
+    @staticmethod
+    def _get_sample_access_logs():
+        """Sample logs collected from the slack api documentation"""
+        return {
```
Comment (ryandeivert) [inline tests/unit/app_integrations/test_apps/test_slack.py]: a bit of a nit, but to be consistent with the codebase: can you ensure all string literals use single quotes (`'`) not double quotes? and omit the leading `u` prefix. there are a few other instances within this file as well.

**Pull request merged**